App/BraveWidgets/TopNewsWidget.swift

```diff
@@ -0,0 +1,241 @@
+// Copyright 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import WidgetKit
+import SwiftUI
+import CodableHelpers
+import SDWebImageSwiftUI
+import Shared
+import DesignSystem
+import os
+import AVFoundation
+import Collections
+
+struct TopNewsWidget: Widget {
+  var supportedFamilies: [WidgetFamily] {
+    if #available(iOS 16.0, *) {
+      return [.systemSmall, .accessoryRectangular]
+    } else {
+      return [.systemSmall]
+    }
+  }
+  
+  var body: some WidgetConfiguration {
+    StaticConfiguration(kind: "TopNewsWidget", provider: TopNewsWidgetProvider()) { entry in
+      TopNewsView(entry: entry)
+    }
+    .supportedFamilies(supportedFamilies)
+    .configurationDisplayName(Strings.Widgets.newsClusteringWidgetTitle)
+    .description(Strings.Widgets.newsClusteringWidgetDescription)
+  }
+}
+
+private struct TopNewsEntry: TimelineEntry {
+  var date: Date
+  var topic: NewsTopic?
+  var image: UIImage?
+  
+  init(date: Date = .now, topic: NewsTopic?, image: UIImage? = nil) {
+    self.date = date
+    self.topic = topic
+    self.image = image
+  }
+}
+
+extension WidgetFamily {
+  var isLockScreen: Bool {
+    switch self {
+    case .accessoryCircular, .accessoryInline, .accessoryRectangular:
+      return true
+    case .systemSmall, .systemMedium, .systemLarge, .systemExtraLarge:
+      return false
+    @unknown default:
+      return false
+    }
+  }
+}
+
+private struct TopNewsWidgetProvider: TimelineProvider {
+  private let model: NewsTopicsModel = .live
+  
+  private func thumbnailSize(for context: Context) -> CGSize {
+    let size = context.displaySize
+    let scale = context.environmentVariants.displayScale?.max() ?? 1.0
+    return .init(width: size.width * scale, height: size.height * scale)
+  }
+  
+  func getSnapshot(in context: Context, completion: @escaping (TopNewsEntry) -> Void) {
```

Comment:
TimelineProvider methods are not async yet or you just prefer it completion handler based?

Reply:
They are not async :(